File: python/stack_03/linked_stack.py
# -*- coding: utf-8 -*-

class Node:
    def __init__(self, value, next_node=None):
        self.value = value
        self.next_node = next_node


class LinkedStack:

    def __init__(self):
        self.top = None

    def push(self, value):
        new_node = Node(value=value)
        new_node.next_node = self.top
        self.top = new_node

    def pop(self):
        if self.top:
            value = self.top.value
            self.top = self.top.next_node
            return value

    def is_empty(self):
        return self.top is None

    def __repr__(self):
        cur = self.top
        values = ""
        while cur:
            values += str(cur.value)
            cur = cur.next_node
            if cur:
                values += '->'
        return values


if __name__ == "__main__":
    stack = LinkedStack()
    for i in range(9):
        stack.push(i)
    print(stack)
    for _ in range(3):
        stack.pop()
    print(stack)

File: go/stack_03/linked_stack.go
package stack_03

import "fmt"

type Node struct {
	val  interface{}
	next *Node
}

type LinkedStack struct {
	topNode *Node
}

func NewLinkedStack() *LinkedStack {
	return &LinkedStack{nil}
}

func (s *LinkedStack) IsEmpty() bool {
	return s.topNode == nil
}

func (s *LinkedStack) Push(value interface{}) {
	fmt.Println("入栈")
	s.topNode = &Node{value, s.topNode}
}

func (s *LinkedStack) Pop() interface{} {
	fmt.Println("出栈")

	if s.IsEmpty() {
		return nil
	}
	v := s.topNode.val
	s.topNode = s.topNode.next
	return v
}

func (s *LinkedStack) Print() {
	if s.IsEmpty() {
		fmt.Println("空栈")
	} else {
		fmt.Println("栈中剩余数据，出栈顺序为：")
		cur := s.topNode
		for cur != nil {
			fmt.Println(cur.val)
			cur = cur.next
		}

	}
}
